src/codegen_envoy.rs

```diff
@@ -370,69 +184,212 @@ impl CodeGen for CodeGenEnvoy {
+            "trace" => make_storage_rpc_value_from_trace(query_data.root_id.clone(), &prop.parent),
+            _ => make_storage_rpc_value_from_target(&prop.parent, &prop.to_dot_string()),
+        },
+        PropertyOrUDF::UdfCall(call) => {
+            // Because of quirky design we need to get the first arg
+            if call.args.len() != 1 {
+                panic!("We currently only implement very specific arguments for UDFs!");
+            }
+            let node = &call.args[0];
+            match node.as_str() {
+                "trace" => make_storage_rpc_value_from_trace(query_data.root_id.clone(), &node),
+                _ => make_storage_rpc_value_from_target(&node, &call.to_ref_str()),
+            }
         }
+    }
+}
 
-        if entity == "trace" {
-            self.make_storage_rpc_value_from_trace(self.ir.root_id.clone(), property);
-        } else {
-            let num_struct_filters = self.ir.struct_filters.len();
-            if !self.ir.struct_filters[num_struct_filters - 1]
-                .vertices
-                .contains(&entity)
-            {
-                panic!("Unknown entity in return expression");
-            }
-            self.make_storage_rpc_value_from_target(entity, property);
+fn make_aggr_block(agg: &Aggregate, query_data: &VisitorResults) -> String {
+    make_return_block(&agg.property, query_data)
+}
+
+fn generate_property_blocks(properties: &IndexSet<Property>) -> Vec<String> {
+    let mut property_blocks = Vec::new();
+    for property in properties {
+        // There is nothing to fetch so ignore.
+        // TODO: What do we actually need here?
+        if property.members.is_empty() {
+            continue;
         }
+        let get_prop_block = format!(
+            "prop_tuple_wrapped = fetch_property(&http_headers.workload_name,
+                                        &{property},
+                                        http_headers);
+                                            ",
+            property = property.as_vec_str(),
+        );
+        property_blocks.push(get_prop_block);
+        let push_block = "if let Some(prop_tuple) = prop_tuple_wrapped {
+            fd.unassigned_properties.push(prop_tuple);
+            } else {
+                return Err(());
+            }"
+        .to_string();
+        property_blocks.push(push_block);
     }
+    property_blocks
+}
 
-    fn make_aggr_block(&mut self) {
-        if self.ir.aggregate.is_none() {
-            return;
+fn generate_udf_blocks(
+    scalar_udf_table: &IndexMap<String, ScalarUdf>,
+    aggregation_udf_table: &IndexMap<String, AggregationUdf>,
+    udf_calls: &IndexSet<UdfCall>,
+) -> Vec<String> {
+    let mut udf_blocks = Vec::new();
+    for call in udf_calls {
+        let udf_ref = call.to_ref_str();
+        if aggregation_udf_table.contains_key(&call.id) {
+            // TODO: Aggregations are handled separately, where do they go?
+            continue;
```

Comment:
Aggregations are handled in the aggregation filter - it looks for entries in aggregation_udf_table to find the correct functions to execute.  This might be a good opportunity to simplify some code - we could do the majority of this within the handlebars through
{{#each scalar_udf_table}} //code in this function {{/each}}

I'm not really sure what the best design choice is here, although we should probably keep it consistent.  I'm leaning towards putting all the code in the handlebars, just because it simplifies our compiler, and our compiler should be focused mostly on whatever information is variable.  Do you have any strong inclination towards keeping auxiliary code (lines 265-305) in the compiler?

Reply:
Not really right now. The reason this line exists is because of avg(), which is parsed as a scalar udf call but the function is actually an aggregation. It can be fixed, but not sure how this should be handled.

Reply:
Are you sure avg is being parsed as a scalar UDF?  If it's in the aggregation_udf_table, and has the headers it has to be parsed, it should be in the aggregation UDF table and parsed as such.  If it's not, something's gone pretty wrong.

Reply:
The parser currently does not know that `avg()` is an aggregation and `height()` is a scalar UDF and so both calls are parsed as generic UDF object. There is no visible difference, even though they are being executed in a radically different fashion. This goes back to the discussion above.

Reply:
Oh, I understand.  I think it should be handled the way it is right now - continuing if it's not a scalar UDF.  I can work on cleaning this function up so that more of the logic is in the handlebars after this pull request.